Source/Shaders/ModelExperimental/MaterialStageFS.glsl

```diff
@@ -0,0 +1,219 @@
+    return vec4(linearOut, srgbIn.a);
+}
+
+vec2 computeTextureTransform(vec2 texCoord, mat3 textureTransform)
+{
+    return vec2(textureTransform * vec3(texCoord, 1.0));
+}
+
+#ifdef HAS_NORMALS
+vec3 computeNormal()
+{
+    vec3 ng = normalize(v_normal);
+
+    vec3 normal = ng;
+    #ifdef HAS_NORMAL_TEXTURE
+    vec2 normalTexCoords = TEXCOORD_NORMAL;
+        #ifdef HAS_NORMAL_TEXTURE_TRANSFORM
+        normalTexCoords = computeTextureTransform(normalTexCoords, u_normalTextureTransform);
+        #endif
+
+        #ifdef HAS_TANGENTS
+        // read tangents from varying
+        vec3 t = normalize(v_tangent.xyz);
+        vec3 b = normalize(cross(ng, t) * v_tangent.w);
+        mat3 tbn = mat3(t, b, ng);
+        vec3 n = texture2D(u_normalTexture, normalTexCoords).rgb;
+        normal = normalize(tbn * (2.0 * n - 1.0));
+        #elif defined(GL_OES_standard_derivatives)
+        // Compute tangents
+        vec3 pos_dx = dFdx(v_positionEC);
+        vec3 pos_dy = dFdy(v_positionEC);
+        vec3 tex_dx = dFdx(vec3(normalTexCoords,0.0));
+        vec3 tex_dy = dFdy(vec3(normalTexCoords,0.0));
+        vec3 t = (tex_dy.t * pos_dx - tex_dx.t * pos_dy) / (tex_dx.s * tex_dy.t - tex_dy.s * tex_dx.t);
+        t = normalize(t - ng * dot(ng, t));
+        vec3 b = normalize(cross(ng, t));
+        mat3 tbn = mat3(t, b, ng);
+        vec3 n = texture2D(u_normalTexture, normalTexCoords).rgb;
+        normal = normalize(tbn * (2.0 * n - 1.0));
+        #endif
+    #endif
+
+    return normal;
+}
+#endif
+
+czm_modelMaterial materialStage(czm_modelMaterial inputMaterial)
+{
+    czm_modelMaterial material = inputMaterial;
+
+    #ifdef HAS_NORMALS
+    material.normal = computeNormal();
+    #endif
+
+    vec4 baseColorWithAlpha = vec4(1.0);
+    // Regardless of whether we use PBR, set a base color
+    #if defined(HAS_BASE_COLOR_TEXTURE)
+    vec2 baseColorTexCoords = TEXCOORD_BASE_COLOR;
+
+        #ifdef HAS_BASE_COLOR_TEXTURE_TRANSFORM
+        baseColorTexCoords = computeTextureTransform(baseColorTexCoords, u_baseColorTextureTransform);
+        #endif
+
+    baseColorWithAlpha = SRGBtoLINEAR4(texture2D(u_baseColorTexture, baseColorTexCoords));
+
+        #ifdef HAS_BASE_COLOR_FACTOR
+        baseColorWithAlpha *= u_baseColorFactor;
+        #endif
+    #elif defined(HAS_BASE_COLOR_FACTOR)
+    baseColorWithAlpha = u_baseColorFactor;
+    #endif
+
+    #ifdef HAS_VERTEX_COLORS
+    baseColorWithAlpha *= v_color_0;
+    #endif
+
+    material.diffuse = baseColorWithAlpha.rgb;
+    material.alpha = baseColorWithAlpha.a;
+
+    #ifdef HAS_OCCLUSION_TEXTURE
+    vec2 occlusionTexCoords = TEXCOORD_OCCLUSION;
+        #ifdef HAS_OCCLUSION_TEXTURE_TRANSFORM
+        occlusionTexCoords = computeTextureTransform(occlusionTexCoords, u_occlusionTextureTransform);
+        #endif
+    material.occlusion = texture2D(u_occlusionTexture, occlusionTexCoords).r;
+    #endif
+
+    #if defined(HAS_EMISSIVE_TEXTURE)
```

Comment:
Why not stick to #ifdef when there's only one condition?

Reply:
Oh okay this was me misunderstanding the `if/elif/else` in GLSL clause, I thought if you had an else clause you can't use `ifdef` but that's not true I'm learning now. You can still use `#ifdef`, there's just no corresponding `#elifdef`, you have to do `#elif defined()`. Here's an example (works in ShaderToy):

```glsl
//#define RED
#define GREEN

void mainImage( out vec4 fragColor, in vec2 fragCoord )
{
    // Normalized pixel coordinates (from 0 to 1)
    vec2 uv = fragCoord/iResolution.xy;

    vec3 color;
    #ifdef RED
    color = vec3(1.0, 0.0, 0.0);
    #elif defined(GREEN)
    color = vec3(0.0, 1.0, 0.0);
    #else
    color= vec3(1.0);
    #endif

    // Output to screen
    fragColor = vec4(color,1.0);
}
```

I'll update the shaders to use `#ifdef` when there's one condition.